tools/build_cleaner.py

```diff
@@ -0,0 +1,463 @@
+"""Find the missing and extra dependencies in Bazel Python build rules.
+
+This program does not modify BUILD files. Instead, it lists the dependencies to
+add and remove.
+
+How to use the build cleaner.
+
+    # Run the following command.
+    tools/build_cleaner.py
+
+If case of a false positive, you might have to update the configuration fields
+BUILD_IN_MODULES, THIRD_PARTY_MODULES, or EXTRA_SOURCE_TO_RULE.
+
+Limitations
+    - The build cleaner only infer the dependencies of py_library rules. For
+        other rules (e.g. py proto, cc wrapper), the mapping should be
+        specified manually in EXTRA_SOURCE_TO_RULE.
+    - In some cases, import of module and import of symbol cannot be
+        differentiated without parsing the python source code. In this case,
+        build cleaner does not try to parse the python source code and rely on a
+        permissive heuristic (possible false negative). See
+        "list_possible_source_of_import" for details.
+"""
+
+from collections import defaultdict
+import ast
+from dataclasses import dataclass
+import re
+import os
+from typing import List, Any, Tuple, Optional, Dict
+
+import colorama
+from colorama import Fore
+
+# Bazel rules to scan for dependencies.
+ALLOWED_BUILD_RULES = {
+    "py_library",
+    "py_binary",
+}
+
+# Python modules that do not require checking / dependencies.
+BUILD_IN_MODULES = {
+    "typing",
+    "os",
+    "abc",
+    "datetime",
+    "math",
+    "csv",
+    "argparse",
+    "collections",
+    "enum",
+    "__future__",
+    "sys",
+    "dataclasses",
+    "time",
+    "io",
+    "google.protobuf.text_format",
+}
+
+# Third party modules that need an "# already_there/" rule.
+#
+# While we assume that third party modules are already installed, we need to
+# list then in comments in the rule "deps" section.
+THIRD_PARTY_MODULES = {
+    "numpy",
+    "pandas",
+    "apache_beam",
+    "tempfile",
+    "logging",
+}
+
+# Mapping from source file to rule.
+#
+# Use EXTRA_SOURCE_TO_RULE to specify mapping that cannot be detected by
+# the build cleaner (e.g. proto, cc code).
+EXTRA_SOURCE_TO_RULE = {
+    ("temporian/proto/core_pb2.py", "//temporian/proto:core_py_proto"),
+    (
+        "temporian/implementation/numpy_cc/operators/operators_cc.py",
+        "//temporian/implementation/numpy_cc/operators:operators_cc",
+    ),
+}
+
+
+# Rule prefix for third party module dependencies.
+THIRD_PARTY_RULE_PREFIX = "# already_there/"
+
+# Filename of BUILD files.
+BUILD_FILENAME = "BUILD"
```

Comment:
BUILD.bazel is an alternative name for the build file and we actually use it for the top-level build file due to a name collision when building the pip package on non case-sensitive systems. Maybe include it here

Reply:
Good point.
*Done